File: app/src/main/java/com/karkoszka/cookingtime/activities/SetPlateActivity.java
package com.karkoszka.cookingtime.activities;

import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;

import com.karkoszka.cookingtime.R;
import com.karkoszka.cookingtime.common.LoaderPreferences;
import com.karkoszka.cookingtime.common.Plate;
import com.karkoszka.cookingtime.fragments.SetTimeFragment;

public class SetPlateActivity
		extends AppCompatActivity
		implements
 SetTimeFragment.OnSetTimeFragmentInteractionListener, 
 SeekBar.OnSeekBarChangeListener {

	static final String PLATE = "plateNumber";
	static final String COLOR = "color";
	static final String HOURS = "hours";
	static final String MINUTES = "minutes";
	static final String SECONDS = "seconds";
	
	private static final int HOURS_MAX = 24;
	private static final int MINUTES_MAX = 59;
	private static final int SECONDS_MAX = 59;
	
	private LoaderPreferences loader;
	private Plate actualPlate;

	
	private int hoursNumber;
	private int minutesNumber;
	private int secondsNumber;
	
	private SeekBar hoursSeekBar;
	private SeekBar minutesSeekBar;
	private SeekBar secondsSeekBar;
	private TextView dynTextHours;
	private TextView dynTextMinutes;
	private TextView dynTextSeconds;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_set_time);
		
		SharedPreferences settings = getSharedPreferences(LoaderPreferences.PREFS_NAME,
				0);
        loader = new LoaderPreferences(settings);
        
        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
	}
	@Override
	public void onResume() {
		super.onResume();		
		int plateNumber = getIntent().getIntExtra(PLATE, 100);
		actualPlate = loader.loadPlate(plateNumber);
		if (getIntent().hasExtra(COLOR)) {
			actualPlate.setColour(getIntent().getIntExtra(COLOR, 100));
		}
		if(getIntent().hasExtra(HOURS)) {
			actualPlate.setHours(getIntent().getIntExtra(HOURS, 0));
			actualPlate.setMinutes(getIntent().getIntExtra(MINUTES, 0));
			actualPlate.setSeconds(getIntent().getIntExtra(SECONDS, 0));
		}
		setTitle("Alarm " + getPlateNumber(plateNumber) );
        
        initControls();
	}

    private String getPlateNumber(int plateNumber) {
	    return Integer.toString(plateNumber + 1);
    }

    @Override
	public boolean onOptionsItemSelected(MenuItem item) {
		if (item.getItemId() == android.R.id.home) {
				onBackPressed();
				return true;
		}
		return super.onOptionsItemSelected(item);
	}

	@Override
	public void onFragmentInteraction(Uri uri) {
		// Just a formality
		
	}
	@Override
	public void onPause() {
		super.onPause();
		save();
	}

	@Override
	public void onBackPressed() {
		save();
		this.finish();
	}
	public void setPlateNumber(View view) {
		save();
	}
	private void save() {
		if (actualPlate.getRuns() == Plate.STARTED &&
				(actualPlate.getHours() != hoursNumber ||
				actualPlate.getMinutes() != minutesNumber ||
				actualPlate.getSeconds() != secondsNumber)) {
			actualPlate.setRuns(Plate.CHANGED);
			Log.d("ST reconfiguring","inner loop" + actualPlate.getId());
		}

		actualPlate.setHours(hoursNumber);
		actualPlate.setMinutes(minutesNumber);
		actualPlate.setSeconds(secondsNumber);
		loader.savePlate(actualPlate);
	}
	public void setColor(View view) {
		Intent intent = new Intent(this, ColorChoserActivity.class);
		intent.putExtra(PLATE, actualPlate.getId());
		intent.putExtra(HOURS, hoursNumber);
		intent.putExtra(MINUTES, minutesNumber);
		intent.putExtra(SECONDS, secondsNumber);
	    startActivity(intent);
	    this.finish();
	}
	private void initControls() {
		dynTextHours = (TextView) findViewById(R.id.dynamicTextHours);
		dynTextMinutes = (TextView) findViewById(R.id.dynamicTextMinutes);
		dynTextSeconds = (TextView) findViewById(R.id.dynamicTextSeconds);
		
		hoursSeekBar = (SeekBar) findViewById(R.id.seekBarHours);
		hoursSeekBar.setOnSeekBarChangeListener(this);
		hoursSeekBar.setMax(HOURS_MAX);
		hoursSeekBar.setProgress(actualPlate.getHours());
		
		minutesSeekBar = (SeekBar) findViewById(R.id.seekBarMinutes);
		minutesSeekBar.setOnSeekBarChangeListener(this);
		minutesSeekBar.setMax(MINUTES_MAX);
		minutesSeekBar.setProgress(actualPlate.getMinutes());
		
		secondsSeekBar = (SeekBar) findViewById(R.id.seekBarSeconds);
		secondsSeekBar.setOnSeekBarChangeListener(this);
		secondsSeekBar.setMax(SECONDS_MAX);
		secondsSeekBar.setProgress(actualPlate.getSeconds());
		
		dynTextHours.setText("" + actualPlate.getHours());
		dynTextMinutes.setText("" + actualPlate.getMinutes());
		dynTextSeconds.setText("" + actualPlate.getSeconds());

		LinearLayout linearSplitSetTime = (LinearLayout) findViewById(R.id.linearSplitSetTime);
		linearSplitSetTime.setBackgroundColor(actualPlate.getColour());
		
	}

	@Override
	public void onProgressChanged(SeekBar seekBar, int progress,
			boolean fromUser) {
		if(seekBar.getId() == hoursSeekBar.getId()) {
			dynTextHours.setText("" + progress);
			hoursNumber = progress;
		}
		else if(seekBar.getId() == minutesSeekBar.getId()) {
			dynTextMinutes.setText("" + progress);
			minutesNumber = progress;
		}
		else if(seekBar.getId() == secondsSeekBar.getId()) {
			dynTextSeconds.setText("" + progress);
			secondsNumber = progress;
		}
	}

	@Override
	public void onStartTrackingTouch(SeekBar seekBar) {
		if(seekBar.getId() == hoursSeekBar.getId()) {
			minutesSeekBar.setEnabled(false);
			secondsSeekBar.setEnabled(false);
		}
		else if(seekBar.getId() == minutesSeekBar.getId()) {
			hoursSeekBar.setEnabled(false);
			secondsSeekBar.setEnabled(false);
		}
		else if(seekBar.getId() == secondsSeekBar.getId()) {
			hoursSeekBar.setEnabled(false);
			minutesSeekBar.setEnabled(false);
		}
	}

	@Override
	public void onStopTrackingTouch(SeekBar seekBar) {
		if(seekBar.getId() == hoursSeekBar.getId()) {
			minutesSeekBar.setEnabled(true);
			secondsSeekBar.setEnabled(true);
		}
		else if(seekBar.getId() == minutesSeekBar.getId()) {
			hoursSeekBar.setEnabled(true);
			secondsSeekBar.setEnabled(true);
		}
		else if(seekBar.getId() == secondsSeekBar.getId()) {
			hoursSeekBar.setEnabled(true);
			minutesSeekBar.setEnabled(true);
		}
		
	}
}

File: app/src/main/java/com/karkoszka/cookingtime/activities/SetPlateActivity.kt
package com.karkoszka.cookingtime.activities

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.*
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentContainerView
import com.karkoszka.cookingtime.R
import com.karkoszka.cookingtime.common.LoaderPreferences
import com.karkoszka.cookingtime.common.OnSwipeTouchListener
import com.karkoszka.cookingtime.common.Plate
import com.karkoszka.cookingtime.fragments.SetTimeFragment.OnSetTimeFragmentInteractionListener

class SetPlateActivity : AppCompatActivity(), OnSetTimeFragmentInteractionListener, OnSeekBarChangeListener {
    private var loader: LoaderPreferences? = null
    private var actualPlate: Plate? = null
    private var hoursNumber = 0
    private var minutesNumber = 0
    private var secondsNumber = 0
    private var hoursSeekBar: SeekBar? = null
    private var minutesSeekBar: SeekBar? = null
    private var secondsSeekBar: SeekBar? = null
    private var dynTextHours: TextView? = null
    private var dynTextMinutes: TextView? = null
    private var dynTextSeconds: TextView? = null
    private lateinit var layout: FragmentContainerView

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_set_time)
        val settings = getSharedPreferences(LoaderPreferences.PREFS_NAME,
                0)
        loader = LoaderPreferences(settings)
        val actionBar = supportActionBar
        actionBar!!.setDisplayHomeAsUpEnabled(true)
        layout = findViewById(R.id.set_time_fragment)
        initSwipeSettings()
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun initSwipeSettings() {
        layout.setOnTouchListener(@SuppressLint("ClickableViewAccessibility")
        object : OnSwipeTouchListener(this@SetPlateActivity) {
            override fun onSwipeLeft() {
                super.onSwipeLeft()
                val intent = Intent(this@SetPlateActivity, SetPlateActivity::class.java)
                intent.putExtra(PLATE, nextPlate(actualPlate!!.id))
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                startActivity(intent)
                overridePendingTransition(R.anim.slide_in_right,
                        R.anim.slide_out_left)
            }
            override fun onSwipeRight() {
                super.onSwipeRight()
                val intent = Intent(this@SetPlateActivity, SetPlateActivity::class.java)
                intent.putExtra(PLATE,  previousPlate(actualPlate!!.id))
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                startActivity(intent)
                overridePendingTransition(R.anim.slide_in_left,
                        R.anim.slide_out_right)
            }
        })
    }

    private fun previousPlate(id: Int): Int {
        if (actualPlate!!.id == 0)
            return 5
        return id - 1
    }

    private fun nextPlate(id: Int): Int {
        if (actualPlate!!.id == 5)
            return 0
        return id + 1
    }

    public override fun onResume() {
        super.onResume()
        val plateNumber = intent.getIntExtra(PLATE, 100)
        actualPlate = loader!!.loadPlate(plateNumber)
        if (intent.hasExtra(COLOR)) {
            actualPlate!!.color = intent.getIntExtra(COLOR, 100)
        }
        if (intent.hasExtra(HOURS)) {
            actualPlate!!.hours = intent.getIntExtra(HOURS, 0)
            actualPlate!!.minutes = intent.getIntExtra(MINUTES, 0)
            actualPlate!!.seconds = intent.getIntExtra(SECONDS, 0)
        }
        title = "Alarm " + getPlateNumber(plateNumber)
        initControls()
    }

    private fun getPlateNumber(plateNumber: Int): String {
        return (plateNumber + 1).toString()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    public override fun onPause() {
        super.onPause()
        save()
    }

    override fun onBackPressed() {
        save()
        finish()
        val intent = Intent(this@SetPlateActivity, MainActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
        startActivity(intent)
    }

    private fun save() {
        if (actualPlate!!.runs == Plate.STARTED &&
                (actualPlate!!.hours != hoursNumber || actualPlate!!.minutes != minutesNumber || actualPlate!!.seconds != secondsNumber)) {
            actualPlate!!.runs = Plate.CHANGED
            Log.d("ST reconfiguring", "inner loop" + actualPlate!!.id)
        }
        actualPlate!!.hours = hoursNumber
        actualPlate!!.minutes = minutesNumber
        actualPlate!!.seconds = secondsNumber
        loader!!.savePlate(actualPlate!!)
    }

    @Suppress("UNUSED_PARAMETER")
    fun setColor(view: View?) {
        val intent = Intent(this, ColorChooserActivity::class.java)
        intent.putExtra(PLATE, actualPlate!!.id)
        intent.putExtra(HOURS, hoursNumber)
        intent.putExtra(MINUTES, minutesNumber)
        intent.putExtra(SECONDS, secondsNumber)
        startActivity(intent)
        finish()
    }

    private fun initControls() {
        dynTextHours = findViewById<View>(R.id.dynamicTextHours) as TextView
        dynTextMinutes = findViewById<View>(R.id.dynamicTextMinutes) as TextView
        dynTextSeconds = findViewById<View>(R.id.dynamicTextSeconds) as TextView
        hoursSeekBar = findViewById<View>(R.id.seekBarHours) as SeekBar
        hoursSeekBar!!.setOnSeekBarChangeListener(this)
        hoursSeekBar!!.max = HOURS_MAX
        hoursSeekBar!!.progress = actualPlate!!.hours
        minutesSeekBar = findViewById<View>(R.id.seekBarMinutes) as SeekBar
        minutesSeekBar!!.setOnSeekBarChangeListener(this)
        minutesSeekBar!!.max = MINUTES_MAX
        minutesSeekBar!!.progress = actualPlate!!.minutes
        secondsSeekBar = findViewById<View>(R.id.seekBarSeconds) as SeekBar
        secondsSeekBar!!.setOnSeekBarChangeListener(this)
        secondsSeekBar!!.max = SECONDS_MAX
        secondsSeekBar!!.progress = actualPlate!!.seconds
        dynTextHours!!.text = getTimeLabel(actualPlate!!.hours)
        dynTextMinutes!!.text = getTimeLabel(actualPlate!!.minutes)
        dynTextSeconds!!.text = getTimeLabel(actualPlate!!.seconds)
        layout.setBackgroundColor(actualPlate!!.color)
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int,
                                   fromUser: Boolean) {
        when (seekBar.id) {
            hoursSeekBar!!.id -> {
                dynTextHours!!.text = getTimeLabel(progress)
                hoursNumber = progress
            }
            minutesSeekBar!!.id -> {
                dynTextMinutes!!.text = getTimeLabel(progress)
                minutesNumber = progress
            }
            secondsSeekBar!!.id -> {
                dynTextSeconds!!.text = getTimeLabel(progress)
                secondsNumber = progress
            }
        }
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {
        when (seekBar.id) {
            hoursSeekBar!!.id -> {
                minutesSeekBar!!.isEnabled = false
                secondsSeekBar!!.isEnabled = false
            }
            minutesSeekBar!!.id -> {
                hoursSeekBar!!.isEnabled = false
                secondsSeekBar!!.isEnabled = false
            }
            secondsSeekBar!!.id -> {
                hoursSeekBar!!.isEnabled = false
                minutesSeekBar!!.isEnabled = false
            }
        }
    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {
        when (seekBar.id) {
            hoursSeekBar!!.id -> {
                minutesSeekBar!!.isEnabled = true
                secondsSeekBar!!.isEnabled = true
            }
            minutesSeekBar!!.id -> {
                hoursSeekBar!!.isEnabled = true
                secondsSeekBar!!.isEnabled = true
            }
            secondsSeekBar!!.id -> {
                hoursSeekBar!!.isEnabled = true
                minutesSeekBar!!.isEnabled = true
            }
        }
    }

    private fun getTimeLabel(progress: Int): String {
        return if (progress < 10) "0$progress" else "" + progress
    }

    companion object {
        const val PLATE = "plateNumber"
        const val COLOR = "color"
        const val HOURS = "hours"
        const val MINUTES = "minutes"
        const val SECONDS = "seconds"
        private const val HOURS_MAX = 24
        private const val MINUTES_MAX = 59
        private const val SECONDS_MAX = 59
    }

    override fun onFragmentInteraction(uri: Uri?) {
        //Just formality
    }
}